src/metabase/db/liquibase.clj

```diff
@@ -181,7 +181,7 @@
 
   IMPORTANT: this function takes `data-source` but not `liquibase` because `.listUnrunChangeSets` is buggy. See #38257."
   [^DataSource data-source]
-  (with-liquibase [liquibase (.getConnection data-source)]
+  (with-liquibase [liquibase data-source]
```

Comment:
See the commit message for details and repro.

Reply:
that looks like a problem with `with-liquibase` tho, maybe you can investigate to see where we leave a stay connection unclosed?

Maybe it's not apparent enough but I have the "IMPORTANT" note above there for a reason :(

Reply:
I think this is the exact place where we leave the conn unclosed. In most other places, `.getConnection` is used with `with-open`, but not here.

I think this fn should either use `with-open`, so the conn is always correctly closed under all circumstances, or it shouldn't open the conn at all and let the called fn do it. I went for the latter option because it looked simpler.

Reply:
ah okay, i just realized that `with-liquibase` will open a connection if it takes a  data-source. all good then!